File: csharp/Assembler/Performance/PerformanceUtils.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;
using Arshu.App.Json;
using Assembler.Engine;
using Assembler.Loader;
using Assembler.Config;
using System.Text.Json.Serialization;

namespace Assembler.Performance;

public static class PerformanceUtils
{
    public class PerfSummaryRow
    {
        [JsonPropertyName("AppSite")]
        public string? AppSite { get; set; }

        [JsonPropertyName("AppFile")]
        public string? AppFile { get; set; }

        [JsonPropertyName("AppView")]
        public string? AppView { get; set; }

        [JsonPropertyName("Iterations")]
        public int Iterations { get; set; }

        [JsonPropertyName("NormalTimeTicks")]
        public long NormalTimeTicks { get; set; }

        [JsonPropertyName("NormalJsonTimeTicks")]
        public long NormalJsonTimeTicks { get; set; }

        [JsonPropertyName("PreProcessTimeTicks")]
        public long PreProcessTimeTicks { get; set; }

        [JsonPropertyName("PreProcessJsonTimeTicks")]
        public long PreProcessJsonTimeTicks { get; set; }

        [JsonPropertyName("OutputSize")]
        public int OutputSize { get; set; }

        [JsonPropertyName("ResultsMatch")]
        public string? ResultsMatch { get; set; }

        [JsonPropertyName("PerfDifference")]
        public string? PerfDifference { get; set; }

        [JsonPropertyName("ScenarioTotalTimeMs")]
        public long ScenarioTotalTimeMs { get; set; }

        [JsonPropertyName("ElapsedTimeMs")]
        public long ElapsedTimeMs { get; set; }

        // Helper properties for display
        [JsonPropertyName("NormalTimeMs")]
        public double NormalTimeMs => (double)NormalTimeTicks / Stopwatch.Frequency * 1000;

        [JsonPropertyName("NormalJsonTimeMs")]
        public double NormalJsonTimeMs => (double)NormalJsonTimeTicks / Stopwatch.Frequency * 1000;

        [JsonPropertyName("PreProcessTimeMs")]
        public double PreProcessTimeMs => (double)PreProcessTimeTicks / Stopwatch.Frequency * 1000;

        [JsonPropertyName("PreProcessJsonTimeMs")]
        public double PreProcessJsonTimeMs => (double)PreProcessJsonTimeTicks / Stopwatch.Frequency * 1000;
    }

    public static List<PerfSummaryRow> RunPerformanceComparison(string assemblerWebDirPath, string projectDirectory, List<Scenario> scenarios, string searchAppSites, bool skipDetails = false, bool enableJsonProcessing = true)
    {
        var startTime = Stopwatch.StartNew();

        if (string.IsNullOrEmpty(assemblerWebDirPath))
        {
            Console.WriteLine("❌ No assemblerWebDirPath passed");
            return new List<PerfSummaryRow>();
        }

        if (scenarios == null || scenarios.Count == 0)
        {
            Console.WriteLine("❌ No scenarios passed");
            return new List<PerfSummaryRow>();
        }

        int iterations = 1000;
        var summaryRows = new List<PerfSummaryRow>();

        foreach (var scenario in scenarios)
        {
            var scenarioStartTime = Stopwatch.StartNew();
            var testAppSite = scenario.AppSite;
            var appFileName = scenario.AppFile;
            var appView = scenario.AppView;
            var appViewPrefix = string.IsNullOrEmpty(appView) ? "" : appView.Substring(0, Math.Min(appView.Length, 6));

            try
            {
                // Clear global caches before loading new templates
                new LoaderNormal().ClearCache();
                new LoaderPreProcess().ClearCache();

                var loaderNormalJson = new LoaderNormalJson(assemblerWebDirPath, testAppSite, searchAppSites);
                var loaderPreProcessJson = new LoaderPreProcessJson(assemblerWebDirPath, testAppSite, searchAppSites);

                if (!skipDetails)
                {
                    Console.WriteLine(new string('-', 60));
                    Console.WriteLine($"[C#] Testing: AppSite={testAppSite}, AppFile={appFileName}, AppView={appView}");
                    Console.WriteLine($"[C#] Iterations: {iterations:N0}");
                }
                var normalJsonEngine = new EngineNormalJson();
                normalJsonEngine.AppViewPrefix = appViewPrefix;

                // JIT Warmup - run a few iterations first to warm up the JIT
                for (int warmup = 0; warmup < 100; warmup++)
                {
                    normalJsonEngine.MergeTemplates(testAppSite, appFileName, appView, loaderNormalJson, enableJsonProcessing);
                }

                var sw = Stopwatch.StartNew();
                string resultNormalJson = "";
                for (int i = 0; i < iterations; i++)
                {
                    resultNormalJson = normalJsonEngine.MergeTemplates(testAppSite, appFileName, appView, loaderNormalJson, enableJsonProcessing);
                }
                sw.Stop();
                var normalJsonTime = sw.ElapsedMilliseconds;
                var normalJsonTicks = sw.ElapsedTicks;
                if (!skipDetails)
                {
                    Console.WriteLine($"[C#] NormalJson Engine:     {normalJsonTime}ms | Avg: {(double)normalJsonTime / iterations:F3}ms/op | Size: {resultNormalJson.Length} chars");
                }
                // Clear global caches before loading new templates
                new LoaderNormal().ClearCache();
                new LoaderPreProcess().ClearCache();
                var preProcessJsonEngine = new EnginePreProcessJson();
                preProcessJsonEngine.AppViewPrefix = appViewPrefix;

                // JIT Warmup for PreProcessJson engine
                for (int warmup = 0; warmup < 100; warmup++)
                {
                    preProcessJsonEngine.MergeTemplates(testAppSite, appFileName, appView, loaderPreProcessJson, enableJsonProcessing);
                }

                sw.Restart();
                string resultPreProcessJson = "";
                for (int i = 0; i < iterations; i++)
                {
                    resultPreProcessJson = preProcessJsonEngine.MergeTemplates(testAppSite, appFileName, appView, loaderPreProcessJson, enableJsonProcessing);
                }
                sw.Stop();
                var preProcessJsonTime = sw.ElapsedMilliseconds;
                var preProcessJsonTicks = sw.ElapsedTicks;
                if (!skipDetails)
                {
                    Console.WriteLine($"[C#] PreProcessJson Engine: {preProcessJsonTime}ms | Avg: {(double)preProcessJsonTime / iterations:F3}ms/op | Size: {resultPreProcessJson.Length} chars");
                    var difference = preProcessJsonTime - normalJsonTime;
                    var differencePercent = normalJsonTime > 0 ? ((double)difference / normalJsonTime) * 100 : 0;
                    Console.WriteLine($"[C#] Performance: {(difference >= 0 ? "+" : "")}{difference}ms ({(differencePercent >= 0 ? "+" : "")}{differencePercent:F1}%) | Match: {(resultNormalJson == resultPreProcessJson ? "YES" : "NO")}");
                }

                // Verify both engines match
                bool allMatch = resultNormalJson == resultPreProcessJson;
                if (!skipDetails)
                {
                    Console.WriteLine($"[C#] Both Engines Match: {(allMatch ? "YES" : "NO")}");
                }

                scenarioStartTime.Stop();
                var scenarioTotalTime = scenarioStartTime.ElapsedMilliseconds;
                var elapsedTime = startTime.ElapsedMilliseconds;

                if (!skipDetails)
                {
                    Console.WriteLine($"[C#] Scenario Total Time: {scenarioTotalTime}ms | Elapsed: {elapsedTime}ms");
                }

                summaryRows.Add(new PerfSummaryRow
                {
                    AppSite = testAppSite,
                    AppFile = appFileName,
                    AppView = appView,
                    Iterations = iterations,
                    NormalTimeTicks = 0, // Old engine removed
                    NormalJsonTimeTicks = normalJsonTicks,
                    PreProcessTimeTicks = 0, // Old engine removed
                    PreProcessJsonTimeTicks = preProcessJsonTicks,
                    OutputSize = resultNormalJson.Length,
                    ResultsMatch = allMatch ? "YES" : "NO",
                    PerfDifference = normalJsonTicks > 0 ? $"{((double)(preProcessJsonTicks - normalJsonTicks) / normalJsonTicks * 100):F1}%" : "0%",
                    ScenarioTotalTimeMs = scenarioTotalTime,
                    ElapsedTimeMs = elapsedTime
                });
            }
            catch (Exception)
            {
                // Silent error handling
            }
        }
        startTime.Stop();

        Console.WriteLine($"\n========== Performance Testing Completed in {startTime.ElapsedMilliseconds}ms ==========\n");

        return summaryRows;
    }

    public static void PrintPerfSummaryTable(string assemblerWebDirPath, string projectDirectory, List<PerfSummaryRow> summaryRows)
    {
        if (summaryRows == null || summaryRows.Count == 0)
            return;
        Console.WriteLine("\n==================== C# PERFORMANCE SUMMARY ====================\n");

        var headers = new[] { "AppSite", "AppView", "Normal(ms)", "NormalJson(ms)", "PreProc(ms)", "PreProcJson(ms)", "Match", "PerfDiff", "ScnTime(ms)", "Elapsed(ms)" };
        int colCount = headers.Length;
        int[] widths = new int[colCount];
        for (int i = 0; i < colCount; i++)
        {
            widths[i] = headers[i].Length;
        }
        foreach (var row in summaryRows)
        {
            widths[0] = Math.Max(widths[0], row.AppSite?.Length ?? 0);
            widths[1] = Math.Max(widths[1], row.AppView?.Length ?? 0);
            widths[2] = Math.Max(widths[2], row.NormalTimeMs.ToString("F2").Length);
            widths[3] = Math.Max(widths[3], row.NormalJsonTimeMs.ToString("F2").Length);
            widths[4] = Math.Max(widths[4], row.PreProcessTimeMs.ToString("F2").Length);
            widths[5] = Math.Max(widths[5], row.PreProcessJsonTimeMs.ToString("F2").Length);
            widths[6] = Math.Max(widths[6], row.ResultsMatch?.Length ?? 0);
            widths[7] = Math.Max(widths[7], row.PerfDifference?.Length ?? 0);
            widths[8] = Math.Max(widths[8], row.ScenarioTotalTimeMs.ToString().Length);
            widths[9] = Math.Max(widths[9], row.ElapsedTimeMs.ToString().Length);
        }
        // Print header
        Console.Write("| ");
        for (int i = 0; i < colCount; i++)
        {
            Console.Write(headers[i].PadRight(widths[i]));
            if (i < colCount - 1) Console.Write(" | ");
        }
        Console.WriteLine(" |");
        // Print divider
        Console.Write("|");
        for (int i = 0; i < colCount; i++)
        {
            Console.Write(" " + new string('-', widths[i]) + " ");
            if (i < colCount - 1) Console.Write("|");
        }
        Console.WriteLine("|");
        // Print rows
        foreach (var row in summaryRows)
        {
            Console.Write("| ");
            Console.Write((row.AppSite ?? "").PadRight(widths[0]));
            Console.Write(" | ");
            Console.Write((row.AppView ?? "").PadRight(widths[1]));
            Console.Write(" | ");
            Console.Write(row.NormalTimeMs.ToString("F2").PadRight(widths[2]));
            Console.Write(" | ");
            Console.Write(row.NormalJsonTimeMs.ToString("F2").PadRight(widths[3]));
            Console.Write(" | ");
            Console.Write(row.PreProcessTimeMs.ToString("F2").PadRight(widths[4]));
            Console.Write(" | ");
            Console.Write(row.PreProcessJsonTimeMs.ToString("F2").PadRight(widths[5]));
            Console.Write(" | ");
            Console.Write((row.ResultsMatch ?? "").PadRight(widths[6]));
            Console.Write(" | ");
            Console.Write((row.PerfDifference ?? "").PadRight(widths[7]));
            Console.Write(" | ");
            Console.Write(row.ScenarioTotalTimeMs.ToString().PadRight(widths[8]));
            Console.Write(" | ");
            Console.Write(row.ElapsedTimeMs.ToString().PadRight(widths[9]));
            Console.WriteLine(" |");
        }
        // Print bottom divider
        Console.Write("|");
        for (int i = 0; i < colCount; i++)
        {
            Console.Write(" " + new string('-', widths[i]) + " ");
            if (i < colCount - 1) Console.Write("|");
        }
        Console.WriteLine("|");

        // Save HTML file
        try
        {
            var html = new StringBuilder();
            html.AppendLine("<!DOCTYPE html>");
            html.AppendLine("<html>");
            html.AppendLine("<head>");
            html.AppendLine("    <meta charset=\"UTF-8\">");
            html.AppendLine("    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">");
            html.AppendLine("    <title>C# Performance Summary Table</title>");
            html.AppendLine("    <style>");
            html.AppendLine("        body { font-family: Arial, sans-serif; margin: 20px; }");
            html.AppendLine("        h2 { color: #333; }");
            html.AppendLine("        .table-container { overflow-x: auto; }");
            html.AppendLine("        table { border-collapse: collapse; width: 100%; min-width: 600px; }");
            html.AppendLine("        th, td { border: 1px solid #ddd; padding: 12px; text-align: left; }");
            html.AppendLine("        th { background-color: #4CAF50; color: white; }");
            html.AppendLine("        tr:nth-child(even) { background-color: #f2f2f2; }");
            html.AppendLine("        @media (max-width: 768px) {");
            html.AppendLine("            body { margin: 10px; }");
            html.AppendLine("            th, td { padding: 8px; font-size: 14px; }");
            html.AppendLine("            h2 { font-size: 20px; }");
            html.AppendLine("        }");
            html.AppendLine("    </style>");
            html.AppendLine("</head>");
            html.AppendLine("<body>");
            html.AppendLine("    <h2>C# Performance Summary Table</h2>");
            html.AppendLine($"    <div class=\"meta\" style=\"color: #666; font-style: italic; margin-bottom: 10px;\">Generated: {DateTime.UtcNow:yyyy-MM-dd HH:mm:ss} UTC | Iterations: 1000, Warmup: 100 | All times in milliseconds (ms)</div>");
            html.AppendLine("    <div class=\"table-container\">");
            html.AppendLine("    <table>");
            html.Append("        <tr>");
            foreach (var h in headers) html.Append($"<th>{h}</th>");
            html.AppendLine("</tr>");
            foreach (var row in summaryRows)
            {
                html.Append("        <tr>");
                html.Append($"<td>{row.AppSite}</td>");
                html.Append($"<td>{row.AppView}</td>");
                html.Append($"<td>{row.NormalTimeMs:F2}</td>");
                html.Append($"<td>{row.NormalJsonTimeMs:F2}</td>");
                html.Append($"<td>{row.PreProcessTimeMs:F2}</td>");
                html.Append($"<td>{row.PreProcessJsonTimeMs:F2}</td>");
                html.Append($"<td>{row.ResultsMatch}</td>");
                html.Append($"<td>{row.PerfDifference}</td>");
                html.Append($"<td>{row.ScenarioTotalTimeMs}</td>");
                html.Append($"<td>{row.ElapsedTimeMs}</td>");
                html.AppendLine("</tr>");
            }
            html.AppendLine("    </table>");
            html.AppendLine("    </div>");
            html.AppendLine("</body>");
            html.AppendLine("</html>");

            var reportsDir = Path.Combine(projectDirectory, "Analysis", "Reports");
            Directory.CreateDirectory(reportsDir);
            var outFile = Path.Combine(reportsDir, "csharp_perfsummary.html");
            File.WriteAllText(outFile, html.ToString());
            Console.WriteLine($"Performance summary HTML saved to: {outFile}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error saving performance summary HTML: {ex.Message}");
        }

        // Save JSON file
        try
        {
            var reportsDir = Path.Combine(projectDirectory, "Analysis", "Reports");
            Directory.CreateDirectory(reportsDir);
            var jsonFile = Path.Combine(reportsDir, "csharp_perfsummary.json");

            // Use custom serialization for NativeAOT compatibility
            var json = SerializePerfSummaryRowsToJson(summaryRows, true);
            File.WriteAllText(jsonFile, json);
            Console.WriteLine($"Performance summary JSON saved to: {jsonFile}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error saving performance summary JSON: {ex.Message}");
        }
    }

    private static string SerializePerfSummaryRowsToJson(List<PerfSummaryRow> rows, bool indented)
    {
        var sb = new StringBuilder();
        sb.Append("[");

        for (int i = 0; i < rows.Count; i++)
        {
            if (i > 0) sb.Append(",");
            if (indented) sb.AppendLine();

            var row = rows[i];
            if (indented) sb.Append("  ");
            sb.Append("{");
            if (indented) sb.AppendLine();

            // Serialize each property manually
            AppendJsonProperty(sb, "AppSite", row.AppSite, indented, false);
            AppendJsonProperty(sb, "AppFile", row.AppFile, indented, false);
            AppendJsonProperty(sb, "AppView", row.AppView, indented, false);
            AppendJsonProperty(sb, "Iterations", row.Iterations.ToString(), indented, false, false);
            AppendJsonProperty(sb, "NormalTimeTicks", row.NormalTimeTicks.ToString(), indented, false, false);
            AppendJsonProperty(sb, "PreProcessTimeTicks", row.PreProcessTimeTicks.ToString(), indented, false, false);
            AppendJsonProperty(sb, "OutputSize", row.OutputSize.ToString(), indented, false, false);
            AppendJsonProperty(sb, "ResultsMatch", row.ResultsMatch, indented, false);
            AppendJsonProperty(sb, "PerfDifference", row.PerfDifference, indented, false);
            AppendJsonProperty(sb, "ScenarioTotalTimeMs", row.ScenarioTotalTimeMs.ToString(), indented, false, false);
            AppendJsonProperty(sb, "ElapsedTimeMs", row.ElapsedTimeMs.ToString(), indented, false, false);
            AppendJsonProperty(sb, "NormalTimeMs", row.NormalTimeMs.ToString("F2"), indented, false, false);
            AppendJsonProperty(sb, "PreProcessTimeMs", row.PreProcessTimeMs.ToString("F2"), indented, true, false);

            if (indented) sb.Append("  ");
            sb.Append("}");
        }

        if (indented) sb.AppendLine();
        sb.Append("]");

        return sb.ToString();
    }

    private static void AppendJsonProperty(StringBuilder sb, string propertyName, string? value, bool indented, bool isLast, bool isString = true)
    {
        if (indented) sb.Append("    ");
        sb.Append("\"");
        sb.Append(propertyName);
        sb.Append("\":");
        if (indented) sb.Append(" ");

        if (isString)
        {
            sb.Append("\"");
            sb.Append(value ?? "");
            sb.Append("\"");
        }
        else
        {
            sb.Append(value ?? "0");
        }

        if (!isLast) sb.Append(",");
        if (indented) sb.AppendLine();
    }
}

File: node/Assembler/src/performance/performanceUtils.js
// Node.js PerformanceUtils - Shared performance comparison and summary table logic
// Matches C#/Rust/Go structure

import { promises as fs } from 'fs';
import fsSync from 'fs';
import path from 'path';
import { LoaderNormal, LoaderPreProcess, EngineNormal, EnginePreProcess } from '../index.js';

export class PerformanceUtils {
    /**
     * Runs performance comparison and returns summary rows
     * @param {string} assemblerWebDirPath
     * @param {Array} scenarios
     * @param {string} searchAppSites
     * @param {boolean} skipDetails
     * @param {boolean} enableJsonProcessing
     * @returns {Array}
     */
    static runPerformanceComparison(assemblerWebDirPath, scenarios, searchAppSites, skipDetails = false, enableJsonProcessing = true) {
        const startTime = Date.now();

        if (!assemblerWebDirPath) {
            console.log("❌ No assemblerWebDirPath passed");
            return [];
        }

        if (!scenarios || scenarios.length === 0) {
            console.log("❌ No scenarios passed");
            return [];
        }

        const iterations = 1000;
        const perfSummaryRows = [];

        for (const scenario of scenarios) {
            const scenarioStartTime = Date.now();
            const testAppSite = scenario.appSite;
            const appFileName = scenario.appFile;
            const appView = scenario.appView;
            const appViewPrefix = appView ? appView.substring(0, Math.min(appView.length, 6)) : "";

            try {
                LoaderNormal.clearCache();
                LoaderPreProcess.clearCache();
                const templates = LoaderNormal.loadGetTemplateFiles(assemblerWebDirPath, testAppSite, searchAppSites);
                const siteTemplates = LoaderPreProcess.loadProcessGetTemplateFiles(assemblerWebDirPath, testAppSite, searchAppSites);
                if (!templates || templates.size === 0)
                    continue;
                const mainTemplateKey = (testAppSite + "_" + appFileName).toLowerCase();
                if (!templates.has(mainTemplateKey))
                    continue;

                if (!skipDetails) {
                    console.log('-'.repeat(60));
                    console.log(`[Node.js] Testing: AppSite=${testAppSite}, AppFile=${appFileName}, AppView=${appView}`);
                    console.log(`[Node.js] Iterations: ${iterations.toLocaleString()}`);
                }
                const normalEngine = new EngineNormal();
                normalEngine.appViewPrefix = appViewPrefix;

                // JIT Warmup - run a few iterations first to warm up the V8 engine
                for (let warmup = 0; warmup < 100; warmup++) {
                    normalEngine.mergeTemplates(testAppSite, appFileName, appView, templates, searchAppSites, enableJsonProcessing);
                }

                const normalStart = process.hrtime.bigint();
                let resultNormal = "";
                for (let i = 0; i < iterations; i++) {
                    resultNormal = normalEngine.mergeTemplates(testAppSite, appFileName, appView, templates, searchAppSites, enableJsonProcessing);
                }
                const normalEnd = process.hrtime.bigint();
                const normalTimeNs = Number(normalEnd - normalStart);
                const normalTimeMs = normalTimeNs / 1_000_000;
                if (!skipDetails) {
                    console.log(`[Node.js] Normal Engine:     ${normalTimeMs.toFixed(0)}ms | Avg: ${(normalTimeMs / iterations).toFixed(3)}ms/op | Size: ${resultNormal.length} chars`);
                }
                LoaderNormal.clearCache();
                LoaderPreProcess.clearCache();
                const preProcessEngine = new EnginePreProcess();
                preProcessEngine.appViewPrefix = appViewPrefix;

                // JIT Warmup for PreProcess engine
                const preprocessedTemplatesObj = Object.fromEntries(siteTemplates.templates);
                for (let warmup = 0; warmup < 100; warmup++) {
                    preProcessEngine.mergeTemplates(testAppSite, appFileName, appView, preprocessedTemplatesObj, searchAppSites, enableJsonProcessing);
                }

                const preProcessStart = process.hrtime.bigint();
                let resultPreProcess = "";
                for (let i = 0; i < iterations; i++) {
                    resultPreProcess = preProcessEngine.mergeTemplates(testAppSite, appFileName, appView, preprocessedTemplatesObj, searchAppSites, enableJsonProcessing);
                }
                const preProcessEnd = process.hrtime.bigint();
                const preProcessTimeNs = Number(preProcessEnd - preProcessStart);
                const preProcessTimeMs = preProcessTimeNs / 1_000_000;
                if (!skipDetails) {
                    console.log(`[Node.js] PreProcess Engine: ${preProcessTimeMs.toFixed(0)}ms | Avg: ${(preProcessTimeMs / iterations).toFixed(3)}ms/op | Size: ${resultPreProcess.length} chars`);
                    const difference = preProcessTimeMs - normalTimeMs;
                    const differencePercent = normalTimeMs > 0 ? (difference / normalTimeMs) * 100 : 0;
                    console.log(`[Node.js] Performance: ${difference >= 0 ? "+" : ""}${difference.toFixed(0)}ms (${differencePercent >= 0 ? "+" : ""}${differencePercent.toFixed(1)}%) | Match: ${resultNormal === resultPreProcess ? "YES" : "NO"}`);
                }

                const scenarioTotalTime = Date.now() - scenarioStartTime;
                const elapsedTime = Date.now() - startTime;

                if (!skipDetails) {
                    console.log(`[Node.js] Scenario Total Time: ${scenarioTotalTime}ms | Elapsed: ${elapsedTime}ms`);
                }

                perfSummaryRows.push({
                    AppSite: testAppSite,
                    AppFile: appFileName,
                    AppView: appView,
                    Iterations: iterations,
                    NormalTimeNanos: normalTimeNs,
                    PreProcessTimeNanos: preProcessTimeNs,
                    NormalTimeMs: normalTimeMs,
                    PreProcessTimeMs: preProcessTimeMs,
                    OutputSize: resultNormal.length,
                    ResultsMatch: (resultNormal === resultPreProcess ? "YES" : "NO"),
                    PerfDifference: normalTimeMs > 0 ? `${((preProcessTimeMs - normalTimeMs) / normalTimeMs * 100).toFixed(1)}%` : "0%",
                    ScenarioTotalTimeMs: scenarioTotalTime,
                    ElapsedTimeMs: elapsedTime
                });
            } catch (error) {
                // Silent error handling
            }
        }

        if (!skipDetails) {
            const elapsed = Date.now() - startTime;
            console.log(`\n========== Performance Testing Completed in ${elapsed}ms ==========\n`);
        }
        return perfSummaryRows;
    }

    /**
     * Prints the performance summary table in markdown format
     * @param {string} assemblerWebDirPath
     * @param {string} projectDirectory
     * @param {Array} summaryRows
     */
    static printPerfSummaryTable(assemblerWebDirPath, projectDirectory, summaryRows) {
        if (!summaryRows || summaryRows.length === 0) {
            return;
        }

        console.log('\n==================== NODE.JS PERFORMANCE SUMMARY ====================\n');

        const headers = ['AppSite', 'AppView', 'Normal(ms)', 'PreProc(ms)', 'Match', 'PerfDiff', 'ScnTime(ms)', 'Elapsed(ms)'];
        const colCount = headers.length;
        const widths = new Array(colCount).fill(0);

        // Calculate column widths
        for (let i = 0; i < colCount; i++) {
            widths[i] = headers[i].length;
        }

        for (const row of summaryRows) {
            const normalMs = (row.NormalTimeNanos || 0) / 1_000_000;
            const preProcessMs = (row.PreProcessTimeNanos || 0) / 1_000_000;
            widths[0] = Math.max(widths[0], (row.AppSite || '').length);
            widths[1] = Math.max(widths[1], (row.AppView || '').length);
            widths[2] = Math.max(widths[2], normalMs.toFixed(2).length);
            widths[3] = Math.max(widths[3], preProcessMs.toFixed(2).length);
            widths[4] = Math.max(widths[4], (row.ResultsMatch || '').length);
            widths[5] = Math.max(widths[5], (row.PerfDifference || '').length);
            widths[6] = Math.max(widths[6], (row.ScenarioTotalTimeMs || 0).toString().length);
            widths[7] = Math.max(widths[7], (row.ElapsedTimeMs || 0).toString().length);
        }

        // Print header
        process.stdout.write('| ');
        for (let i = 0; i < colCount; i++) {
            process.stdout.write(headers[i].padEnd(widths[i]));
            if (i < colCount - 1) process.stdout.write(' | ');
        }
        console.log(' |');

        // Print divider
        process.stdout.write('|');
        for (let i = 0; i < colCount; i++) {
            process.stdout.write(' ' + '-'.repeat(widths[i]) + ' ');
            if (i < colCount - 1) process.stdout.write('|');
        }
        console.log('|');

        // Print rows
        for (const row of summaryRows) {
            const normalMs = (row.NormalTimeNanos || 0) / 1_000_000;
            const preProcessMs = (row.PreProcessTimeNanos || 0) / 1_000_000;
            process.stdout.write('| ');
            process.stdout.write((row.AppSite || '').padEnd(widths[0]));
            process.stdout.write(' | ');
            process.stdout.write((row.AppView || '').padEnd(widths[1]));
            process.stdout.write(' | ');
            process.stdout.write(normalMs.toFixed(2).padEnd(widths[2]));
            process.stdout.write(' | ');
            process.stdout.write(preProcessMs.toFixed(2).padEnd(widths[3]));
            process.stdout.write(' | ');
            process.stdout.write((row.ResultsMatch || '').padEnd(widths[4]));
            process.stdout.write(' | ');
            process.stdout.write((row.PerfDifference || '').padEnd(widths[5]));
            process.stdout.write(' | ');
            process.stdout.write((row.ScenarioTotalTimeMs || 0).toString().padEnd(widths[6]));
            process.stdout.write(' | ');
            process.stdout.write((row.ElapsedTimeMs || 0).toString().padEnd(widths[7]));
            console.log(' |');
        }

        // Print bottom divider
        process.stdout.write('|');
        for (let i = 0; i < colCount; i++) {
            process.stdout.write(' ' + '-'.repeat(widths[i]) + ' ');
            if (i < colCount - 1) process.stdout.write('|');
        }
        console.log('|');

        // Save HTML file
        try {
            const reportsDir = path.join(projectDirectory, 'Analysis', 'Reports');
            if (!fsSync.existsSync(reportsDir)) {
                fsSync.mkdirSync(reportsDir, { recursive: true });
            }

            const html = [];
            html.push('<!DOCTYPE html>');
            html.push('<html>');
            html.push('<head>');
            html.push('    <meta charset="UTF-8">');
            html.push('    <meta name="viewport" content="width=device-width, initial-scale=1.0">');
            html.push('    <title>Node.js Performance Summary Table</title>');
            html.push('    <style>');
            html.push('        body { font-family: Arial, sans-serif; margin: 20px; }');
            html.push('        h2 { color: #333; }');
            html.push('        .table-container { overflow-x: auto; }');
            html.push('        table { border-collapse: collapse; width: 100%; min-width: 600px; }');
            html.push('        th, td { border: 1px solid #ddd; padding: 12px; text-align: left; }');
            html.push('        th { background-color: #4CAF50; color: white; }');
            html.push('        tr:nth-child(even) { background-color: #f2f2f2; }');
            html.push('        @media (max-width: 768px) {');
            html.push('            body { margin: 10px; }');
            html.push('            th, td { padding: 8px; font-size: 14px; }');
            html.push('            h2 { font-size: 20px; }');
            html.push('        }');
            html.push('    </style>');
            html.push('</head>');
            html.push('<body>');
            html.push('    <h2>Node.js Performance Summary Table</h2>');
            html.push(`    <div class="meta" style="color: #666; font-style: italic; margin-bottom: 10px;">Generated: ${new Date().toISOString().replace('T', ' ').substring(0, 19)} UTC | Iterations: 1000, Warmup: 100 | All times in milliseconds (ms)</div>`);
            html.push('    <div class="table-container">');
            html.push('    <table>');
            html.push('        <tr><th>AppSite</th><th>AppView</th><th>Normal(ms)</th><th>PreProc(ms)</th><th>Match</th><th>PerfDiff</th><th>ScnTime(ms)</th><th>Elapsed(ms)</th></tr>');

            for (const row of summaryRows) {
                const normalMs = (row.NormalTimeNanos || 0) / 1_000_000;
                const preProcessMs = (row.PreProcessTimeNanos || 0) / 1_000_000;
                html.push('        <tr>');
                html.push(`<td>${row.AppSite || ''}</td>`);
                html.push(`<td>${row.AppView || ''}</td>`);
                html.push(`<td>${normalMs.toFixed(2)}</td>`);
                html.push(`<td>${preProcessMs.toFixed(2)}</td>`);
                html.push(`<td>${row.ResultsMatch || ''}</td>`);
                html.push(`<td>${row.PerfDifference || ''}</td>`);
                html.push(`<td>${row.ScenarioTotalTimeMs || 0}</td>`);
                html.push(`<td>${row.ElapsedTimeMs || 0}</td>`);
                html.push('</tr>');
            }

            html.push('    </table>');
            html.push('    </div>');
            html.push('</body>');
            html.push('</html>');

            const outFile = path.join(reportsDir, 'nodejs_perfsummary.html');
            fsSync.writeFileSync(outFile, html.join('\n'), 'utf8');
            console.log(`Performance summary HTML saved to: ${outFile}`);
        } catch (error) {
            console.error(`Error saving performance summary HTML: ${error.message}`);
        }

        // Save JSON file
        try {
            const reportsDir = path.join(projectDirectory, 'Analysis', 'Reports');
            if (!fsSync.existsSync(reportsDir)) {
                fsSync.mkdirSync(reportsDir, { recursive: true });
            }

            const jsonFile = path.join(reportsDir, 'nodejs_perfsummary.json');
            const jsonData = JSON.stringify(summaryRows, null, 2);
            fsSync.writeFileSync(jsonFile, jsonData, 'utf8');
            console.log(`Performance summary JSON saved to: ${jsonFile}`);
        } catch (error) {
            console.error(`Error saving performance summary JSON: ${error.message}`);
        }
    }
}
